src/core/desktopentrymonitor.cpp

```diff
@@ -0,0 +1,57 @@
+#include "desktopentrymonitor.hpp"
+
+#include <qdir.h>
+#include <qdiriterator.h>
+#include <qfileinfo.h>
+#include <qloggingcategory.h>
+
+#include "desktopentry.hpp"
+
+static void addPathAndParents(QFileSystemWatcher& watcher, const QString& path) {
+	watcher.addPath(path);
+
+	QString p = QFileInfo(path).absolutePath();
+	while (!p.isEmpty()) {
+		watcher.addPath(p);
+		const auto parent = QFileInfo(p).dir().absolutePath();
+		if (parent == p) break;
+		p = parent;
+	}
+}
+
+DesktopEntryMonitor::DesktopEntryMonitor(QObject* parent): QObject(parent) {
+	this->debounceTimer.setSingleShot(true);
+	this->debounceTimer.setInterval(100);
+
+	connect(
+	    &this->watcher,
+	    &QFileSystemWatcher::directoryChanged,
+	    this,
+	    &DesktopEntryMonitor::onDirectoryChanged
+	);
+	connect(&this->debounceTimer, &QTimer::timeout, this, &DesktopEntryMonitor::processChanges);
+
+	this->startMonitoring();
+}
+
+void DesktopEntryMonitor::startMonitoring() {
+	for (const auto& path: DesktopEntryManager::desktopPaths()) {
+		if (!QDir(path).exists()) continue;
+		addPathAndParents(this->watcher, path);
+		this->scanAndWatch(path);
+	}
+}
+
+void DesktopEntryMonitor::scanAndWatch(const QString& dirPath) {
+	auto dir = QDir(dirPath);
+	if (!dir.exists()) return;
+
+	this->watcher.addPath(dirPath);
+
+	auto subdirs = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot | QDir::NoSymLinks);
+	for (const auto& subdir: subdirs) this->watcher.addPath(subdir.absoluteFilePath());
+}
+
+void DesktopEntryMonitor::onDirectoryChanged(const QString&) { this->debounceTimer.start(); }
```

Comment:
Unfortunately once we include parents it becomes more complex than this. Ideally we don't want to trigger a rescan if a user creates a file in / or something. However this is acceptable for now because fixing it properly will require rewriting QFileSystemWatcher.